File: dimos/navigation/local_planner/obstacles.py
# Copyright 2026 Dimensional Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Which returns are obstacles: a property of the body, not of the scene.

The planner's world is a z-slice of the cloud, and where that slice belongs is
something the robot already knows: the base rides `base_height` above the
surface its feet stand on. Re-reference the cloud to that surface and the
geometry reads itself: below `steppable` the legs negotiate it, above
`height` the body passes underneath, in between it is a wall. Nothing is
estimated off the scene, so nothing can be estimated wrong.

A model is z-only and says nothing about xy. The frame is the caller's: the
adapter re-references the cloud (`adapter/planner.py`) before handing it over,
and `hard_points` is where the two meet.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, field as dc_field
from typing import TYPE_CHECKING, Protocol

import numpy as np
from scipy.spatial import cKDTree

from dimos.navigation.local_planner.search.se2 import VOXEL

if TYPE_CHECKING:
    from numpy.typing import NDArray

    from dimos.navigation.embodiment.base import Embodiment

# Ground exclusion for the body-referenced band: THREE voxel layers. A floor
# whose true height sits near a voxel boundary quantises into the layers either
# side of it, and under a lidar pitched 60 deg down the grazing returns beyond
# a metre read a layer high again (measured on the Go2: floor voxels at +0.02
# under the body, +0.10 at 1-6 m). Their noise then reaches the third layer,
# and at two layers that layer stood as a carpet the search could not cross:
# 18 of 36 waypoints of an open-room route at zero clearance, 0 of 36 at three.
LOW = 3 * VOXEL


def _no_soft() -> NDArray[np.float32]:
    return np.empty((0, 2), dtype=np.float32)


@dataclass(frozen=True)
class ObstacleField:
    """What a model made of one cloud, as indices into that cloud."""

    hard: NDArray[np.int32]  # (K,) never traversable
    soft: NDArray[np.float32] = dc_field(default_factory=_no_soft)  # (M, 2) [index, cost]


class ObstacleModel(Protocol):
    """A z-rule over a cloud referenced to the surface the feet stand on."""

    def field(self, cloud: NDArray[np.float32]) -> ObstacleField: ...


class BodyBand:
    """Obstacles by the body's own geometry: clear of the ground, under the belly."""

    def __init__(self, emb: Embodiment) -> None:
        self.low = LOW
        self.high = emb.height

    def field(self, cloud: NDArray[np.float32]) -> ObstacleField:
        z = cloud[:, 2]
        hard = np.flatnonzero((z > self.low) & (z <= self.high))
        return ObstacleField(hard=hard.astype(np.int32))


OBSTACLE_MODELS: dict[str, Callable[[Embodiment], ObstacleModel]] = {
    "body_band": BodyBand,  # z-referenced zones from the embodiment
}


def load(name: str, emb: Embodiment) -> ObstacleModel:
    """The named model, built for this body."""
    factory = OBSTACLE_MODELS.get(name)
    if factory is None:
        raise ValueError(f"unknown obstacle model {name!r}; known: {sorted(OBSTACLE_MODELS)}")
    return factory(emb)


def referenced(cloud: NDArray[np.float32], ground_z: float) -> NDArray[np.float32]:
    """The cloud in the frame a model reads: z off the support surface, finite rows only."""
    pts = np.asarray(cloud, dtype=np.float32).reshape(-1, 3)
    pts = pts[np.isfinite(pts).all(axis=1)]
    return pts - np.array([0.0, 0.0, ground_z], dtype=np.float32)


def ground_points(cloud: NDArray[np.float32], ground_z: float) -> NDArray[np.float64]:
    """The floor the cloud saw, as (N, 2) xy: returns at or under the band."""
    pts = referenced(cloud, ground_z)
    return np.ascontiguousarray(pts[pts[:, 2] <= LOW][:, :2], dtype=np.float64)


def hard_points(
    model: ObstacleModel, cloud: NDArray[np.float32], ground_z: float
) -> NDArray[np.float32]:
    """The obstacles this model sees: the cloud the search plans on."""
    pts = referenced(cloud, ground_z)
    return np.ascontiguousarray(pts[model.field(pts).hard])


def path_clearance(
    xy: NDArray[np.float64], points: NDArray[np.float32], half_width: float
) -> NDArray[np.float64]:
    """Per-waypoint room hint (m): nearest obstacle minus the half-width.

    A speed hint for the controller, not a safety contract. Nothing to hit or
    an empty path = infinite room. `points` is a model's hard set: every row is
    something the body can hit, and z rides along unread; deciding that again
    here would price a different world than the plan was made for.
    """
    xy = np.asarray(xy, dtype=np.float64).reshape(-1, 2)
    band = np.asarray(points, dtype=np.float32).reshape(-1, 3)[:, :2]
    if not len(band) or not len(xy):
        return np.full(len(xy), np.inf)
    d, _ = cKDTree(band).query(xy)
    return np.asarray(d, dtype=np.float64) - half_width

File: dimos/navigation/local_planner/rust/src/module/obstacles.rs
// Copyright 2026 Dimensional Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Which returns are obstacles: the rust twin of `motion/obstacles.py`, which is the specification.
//! A model is z-only over a cloud referenced to the surface the feet stand on; [`hard_points`] joins it to the caller's frame.

use crate::planner::Emb;

/// Ground exclusion (m): three voxel layers, since the floor reads a layer high at range (see obstacles.py).
pub const LOW: f64 = 0.24;

/// The model names a config may carry, for the validation error message.
pub const MODELS: [&str; 1] = ["body_band"];

/// What a model made of one cloud, as indices into that cloud.
#[derive(Debug, Default, PartialEq)]
pub struct Field {
    /// Never traversable.
    pub hard: Vec<u32>,
    /// Traversable at a price: `(index, cost)`. Empty for now.
    pub soft: Vec<(u32, f32)>,
}

/// A z-rule over a cloud referenced to the surface the feet stand on.
pub trait ObstacleModel: Send + Sync {
    fn field(&self, cloud: &[[f32; 3]]) -> Field;
}

/// Obstacles by the body's own geometry: clear of the ground, under the belly.
pub struct BodyBand {
    low: f32,
    high: f32,
}

impl ObstacleModel for BodyBand {
    fn field(&self, cloud: &[[f32; 3]]) -> Field {
        Field {
            hard: indices(cloud, |z| z > self.low && z <= self.high),
            soft: Vec::new(),
        }
    }
}

fn indices(cloud: &[[f32; 3]], keep: impl Fn(f32) -> bool) -> Vec<u32> {
    cloud
        .iter()
        .enumerate()
        .filter(|(_, p)| keep(p[2]))
        .map(|(i, _)| i as u32)
        .collect()
}

/// The named model, built for this body, or `None` when the name is unknown.
pub fn load(name: &str, emb: &Emb) -> Option<Box<dyn ObstacleModel>> {
    match name {
        "body_band" => Some(Box::new(BodyBand {
            low: LOW as f32,
            high: emb.height as f32,
        })),
        _ => None,
    }
}

/// z off the support surface; f32 throughout, as the python does.
pub fn referenced(points: &[[f32; 3]], ground_z: f64) -> Vec<[f32; 3]> {
    let ground = ground_z as f32;
    points.iter().map(|p| [p[0], p[1], p[2] - ground]).collect()
}

/// The floor the cloud saw, as xy; lattice cells with none are priced as unseen (`Config::unseen_cost`).
pub fn ground_points(points: &[[f32; 3]], ground_z: f64) -> Vec<[f64; 2]> {
    let ground = ground_z as f32;
    points
        .iter()
        .filter(|p| p[2] - ground <= LOW as f32)
        .map(|p| [p[0] as f64, p[1] as f64])
        .collect()
}

/// The obstacles this model sees: the cloud the search plans on.
pub fn hard_points(model: &dyn ObstacleModel, points: &[[f32; 3]], ground_z: f64) -> Vec<[f32; 3]> {
    let pts = referenced(points, ground_z);
    model
        .field(&pts)
        .hard
        .iter()
        .map(|&i| pts[i as usize])
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn fixture() -> Emb {
        Emb::fixture()
    }

    /// A ground slab under an obstacle just above LOW, lifted to `ground_z`.
    fn room(ground_z: f32) -> Vec<[f32; 3]> {
        let mut pts = Vec::new();
        for x in [-1.0f32, 0.0, 1.0] {
            for z in [0.0f32, 0.04, 0.08, 0.12] {
                pts.push([x, 0.0, z + ground_z]);
            }
        }
        for z in obstacle_heights() {
            pts.push([2.0, 0.0, z + ground_z]);
        }
        pts
    }

    /// Three heights just above the ground exclusion, so the fixture follows LOW.
    fn obstacle_heights() -> [f32; 3] {
        let low = LOW as f32;
        [low + 0.02, low + 0.08, low + 0.14]
    }

    #[test]
    fn the_body_band_drops_the_ground_slab_and_keeps_the_obstacle() {
        // the phantom regression: a quantised floor must not become a wall
        let model = load("body_band", &fixture()).expect("known model");
        let got = hard_points(model.as_ref(), &room(-0.28), -0.28);
        assert_eq!(got.len(), 3, "{got:?}");
        for (p, want) in got.iter().zip(obstacle_heights()) {
            assert!((p[2] - want).abs() < 1e-6, "{p:?} wanted {want}");
        }
    }

    #[test]
    fn the_body_band_looks_under_the_belly_not_over_it() {
        let model = load("body_band", &fixture()).expect("known model");
        let cloud = [[1.0, 0.0, 0.3], [1.0, 0.0, 0.46]];
        let got = hard_points(model.as_ref(), &cloud, 0.0);
        assert_eq!(got, vec![[1.0, 0.0, 0.3]]);
    }

    #[test]
    fn an_unknown_model_is_refused_rather_than_defaulted() {
        assert!(load("floor_anchor", &fixture()).is_none());
        for name in MODELS {
            assert!(
                load(name, &fixture()).is_some(),
                "{name} is advertised but unknown"
            );
        }
    }

    #[test]
    fn the_soft_tier_is_empty_for_now() {
        let model = load("body_band", &fixture()).expect("known model");
        assert!(model.field(&room(0.0)).soft.is_empty());
    }
}
